File: src/semantic/layer.ts
/**
 * Semantic layer — YAML read/write for table semantics.
 * Stores table-level business meaning as YAML files under semantic_store/{source_id}/{table}.yaml
 */

import fs from "node:fs";
import path from "node:path";
import yaml from "js-yaml";
import type { TableSemantic } from "../types.js";
import { getConfig } from "../config.js";

/**
 * Sanitize a source ID for use as a directory name.
 * "mysql:mydb" → "mysql_mydb", "csv:orders.csv" → "csv_orders_csv"
 */
function sanitizeSourceId(sourceId: string): string {
  return sourceId.replace(/[^a-zA-Z0-9_\u4e00-\u9fff-]/g, "_");
}

/**
 * Get the file path for a table's semantic YAML.
 */
function getSemanticFilePath(sourceId: string, tableName: string): string {
  const config = getConfig();
  const dir = path.join(config.semantic.storePath, sanitizeSourceId(sourceId));
  return path.join(dir, `${tableName}.yaml`);
}

/**
 * Read a table's semantic definition from YAML.
 * Returns null if the file doesn't exist.
 */
export function readTableSemantic(
  sourceId: string,
  tableName: string,
): TableSemantic | null {
  const filePath = getSemanticFilePath(sourceId, tableName);
  if (!fs.existsSync(filePath)) {
    return null;
  }
  try {
    const content = fs.readFileSync(filePath, "utf-8");
    return yaml.load(content) as TableSemantic;
  } catch {
    return null;
  }
}

/**
 * Write a table's semantic definition to YAML.
 * Creates directory structure if needed.
 */
export function writeTableSemantic(
  sourceId: string,
  tableName: string,
  semantic: TableSemantic,
): void {
  const filePath = getSemanticFilePath(sourceId, tableName);
  const dir = path.dirname(filePath);
  fs.mkdirSync(dir, { recursive: true });

  // Ensure updatedAt is set
  semantic.updatedAt = new Date().toISOString();

  const content = yaml.dump(semantic, {
    indent: 2,
    lineWidth: 120,
    noRefs: true,
    sortKeys: false,
  });

  fs.writeFileSync(filePath, content, "utf-8");
}

/**
 * List all semantic definitions for a source.
 */
export function listSemantics(sourceId: string): TableSemantic[] {
  const config = getConfig();
  const dir = path.join(config.semantic.storePath, sanitizeSourceId(sourceId));

  if (!fs.existsSync(dir)) {
    return [];
  }

  const files = fs.readdirSync(dir).filter((f) => f.endsWith(".yaml"));
  const results: TableSemantic[] = [];

  for (const file of files) {
    try {
      const content = fs.readFileSync(path.join(dir, file), "utf-8");
      const semantic = yaml.load(content) as TableSemantic;
      if (semantic) results.push(semantic);
    } catch {
      // Skip invalid files
    }
  }

  return results;
}

/**
 * List all semantic definitions across all sources.
 */
export function listAllSemantics(): TableSemantic[] {
  const config = getConfig();
  const storePath = config.semantic.storePath;

  if (!fs.existsSync(storePath)) {
    return [];
  }

  const results: TableSemantic[] = [];
  const sourceDirs = fs
    .readdirSync(storePath, { withFileTypes: true })
    .filter((d) => d.isDirectory());

  for (const dir of sourceDirs) {
    const fullDir = path.join(storePath, dir.name);
    const files = fs.readdirSync(fullDir).filter((f) => f.endsWith(".yaml"));
    for (const file of files) {
      try {
        const content = fs.readFileSync(path.join(fullDir, file), "utf-8");
        const semantic = yaml.load(content) as TableSemantic;
        if (semantic) results.push(semantic);
      } catch {
        // Skip invalid files
      }
    }
  }

  return results;
}

/**
 * Delete a table's semantic definition.
 */
export function deleteTableSemantic(
  sourceId: string,
  tableName: string,
): boolean {
  const filePath = getSemanticFilePath(sourceId, tableName);
  if (fs.existsSync(filePath)) {
    fs.unlinkSync(filePath);
    return true;
  }
  return false;
}

File: lucid_skill/semantic/layer.py
"""YAML-based semantic layer for reading/writing table semantics."""

from __future__ import annotations

import re
from dataclasses import asdict, fields
from datetime import datetime, timezone
from pathlib import Path

import yaml

from lucid_skill.config import get_config
from lucid_skill.types import (
    ColumnSemantic,
    MetricDefinition,
    RelationSemantic,
    TableSemantic,
)


def sanitize_source_id(source_id: str) -> str:
    """Replace non-alphanumeric/underscore/chinese/hyphen characters with '_'."""
    return re.sub(r"[^\w\u4e00-\u9fff\-]", "_", source_id)


def get_semantic_file_path(source_id: str, table_name: str) -> Path:
    """Return the path to the YAML file for a given source and table."""
    config = get_config()
    store_path = Path(config.semantic.store_path)
    return store_path / sanitize_source_id(source_id) / f"{table_name}.yaml"


def _dict_to_table_semantic(data: dict) -> TableSemantic:
    """Convert a raw dict (from YAML) to a TableSemantic dataclass."""
    columns = [
        ColumnSemantic(**col) for col in (data.get("columns") or [])
    ]
    relations = [
        RelationSemantic(**rel) for rel in (data.get("relations") or [])
    ] if data.get("relations") else None
    metrics = [
        MetricDefinition(**m) for m in (data.get("metrics") or [])
    ] if data.get("metrics") else None

    return TableSemantic(
        source=data.get("source", ""),
        table=data.get("table", ""),
        confirmed=data.get("confirmed", False),
        updated_at=data.get("updated_at", ""),
        columns=columns,
        description=data.get("description"),
        business_domain=data.get("business_domain"),
        tags=data.get("tags"),
        relations=relations,
        metrics=metrics,
    )


def _table_semantic_to_dict(semantic: TableSemantic) -> dict:
    """Convert a TableSemantic dataclass to a dict suitable for YAML output."""
    data = asdict(semantic)
    # Remove None values for cleaner YAML
    return {k: v for k, v in data.items() if v is not None}


def read_table_semantic(source_id: str, table_name: str) -> TableSemantic | None:
    """Read a TableSemantic from the YAML file. Returns None if not found."""
    path = get_semantic_file_path(source_id, table_name)
    if not path.exists():
        return None
    try:
        with open(path, "r", encoding="utf-8") as f:
            data = yaml.safe_load(f)
        if not data:
            return None
        return _dict_to_table_semantic(data)
    except Exception:
        return None


def write_table_semantic(source_id: str, table_name: str, semantic: TableSemantic) -> None:
    """Write a TableSemantic to the YAML file, creating directories as needed."""
    path = get_semantic_file_path(source_id, table_name)
    path.parent.mkdir(parents=True, exist_ok=True)

    # Update the timestamp
    semantic.updated_at = datetime.now(timezone.utc).isoformat()

    data = _table_semantic_to_dict(semantic)
    with open(path, "w", encoding="utf-8") as f:
        yaml.dump(data, f, default_flow_style=False, allow_unicode=True, sort_keys=False)


def list_semantics(source_id: str) -> list[TableSemantic]:
    """List all TableSemantic entries for a given source."""
    config = get_config()
    source_dir = Path(config.semantic.store_path) / sanitize_source_id(source_id)
    if not source_dir.exists():
        return []
    results = []
    for yaml_file in sorted(source_dir.glob("*.yaml")):
        try:
            with open(yaml_file, "r", encoding="utf-8") as f:
                data = yaml.safe_load(f)
            if data:
                results.append(_dict_to_table_semantic(data))
        except Exception:
            continue
    return results


def list_all_semantics() -> list[TableSemantic]:
    """List all TableSemantic entries across all sources."""
    config = get_config()
    store_path = Path(config.semantic.store_path)
    if not store_path.exists():
        return []
    results = []
    for source_dir in sorted(store_path.iterdir()):
        if not source_dir.is_dir():
            continue
        for yaml_file in sorted(source_dir.glob("*.yaml")):
            try:
                with open(yaml_file, "r", encoding="utf-8") as f:
                    data = yaml.safe_load(f)
                if data:
                    results.append(_dict_to_table_semantic(data))
            except Exception:
                continue
    return results


def delete_table_semantic(source_id: str, table_name: str) -> bool:
    """Delete a TableSemantic YAML file. Returns True if deleted, False if not found."""
    path = get_semantic_file_path(source_id, table_name)
    if not path.exists():
        return False
    path.unlink()
    return True
